docs/design/resource-qos.md

```diff
@@ -0,0 +1,254 @@
+				memory: 1Gi
+			requests:
+				cpu: 10m
+				memory: 1Gi
+
+	name: bar
+```
+
+Container `foo` and `bar` have limits set for different resources.
+
+```yaml
+containers:
+	name: foo
+		resources:
+			limits:
+				memory: 1Gi
+
+	name: bar
+		resources:
+			limits:
+				cpu: 100m
+```
+
+Container `foo` has no limits set, and `bar` has neither requests nor limits specified.
+
+```yaml
+containers:
+	name: foo
+		resources:
+			requests:
+				cpu: 10m
+				memory: 1Gi
+
+	name: bar
+```
+
+- If `requests` and `limits` are not set for all of the resources, across all containers, then the pod is classified as **Best-Effort**.
+
+Examples:
+
+```yaml
+containers:
+	name: foo
+		resources:
+	name: bar
+		resources:
+```
+
+Pods will not be killed if CPU guarantees cannot be met (for example if system tasks or daemons take up lots of CPU), they will be temporarily throttled.
+
+Memory is an incompressible resource and so let's discuss the semantics of memory management a bit.
+
+- *Best-Effort* pods will be treated as lowest priority. Processes in these pods are the first to get killed if the system runs out of memory.
+These containers can use any amount of free memory in the node though.
+
+- *Guaranteed* pods are considered top-priority and are guaranteed to not be killed until they exceed their limits, or if the system is under memory pressure and there are no lower priority containers that can be evicted.
+
+- *Burstable* pods have some form of minimal resource guarantee, but can use more resources when available.
+Under system memory pressure, these containers are more likely to be killed once they exceed their requests and no *Best-Effort* pods exist.
+
+
+### Alternative QoS Class Policy
+
+An alternative is to have user-specified numerical priorities that guide Kubelet on which tasks to kill (if the node runs out of memory, lower priority tasks will be killed).
+A strict hierarchy of user-specified numerical priorities is not desirable because:
+
+1. Achieved behavior would be emergent based on how users assigned priorities to their pods. No particular SLO could be delivered by the system, and usage would be subject to gaming if not restricted administratively
+2. Changes to desired priority bands would require changes to all user pod configurations.
+
+### OOM Score configuration at the Nodes
+
+Pod OOM score configuration
+- Note that the OOM score of a process is 10 times the % of memory the process consumes, adjusted by OOM_SCORE_ADJ, barring exceptions (e.g. process is launched by root). Processes with higher OOM scores are killed.
+- The base OOM score is between 0 and 1000, so if process A’s OOM_SCORE_ADJ - process B’s OOM_SCORE_ADJ is over a 1000, then process A will always be OOM killed before B.
+- The final OOM score of a process is also between 0 and 1000
+
+*Best-effort*
+	- Set OOM_SCORE_ADJ: 1000
+	- So processes in best-effort containers will have an OOM_SCORE of 1000
+
+*Guaranteed*
+	- Set OOM_SCORE_ADJ: -998
```

Comment:
as noted above, we will need to update the code so that we only do this when pod qos: guaranteed, not just pod qos:besteffort with a memory limit specified.